mosec/manager.py

```diff
@@ -0,0 +1,310 @@
+
+    def append(self, runtime: WorkerRuntime):
+        """Sequentially appends workers to the workflow pipeline."""
+        self._worker_runtimes.append(runtime)
+
+    def _label_stage(self, stage_id: int):
+        stage = ""
+        if stage_id == 0:
+            stage += STAGE_INGRESS
+        if stage_id == self.worker_count - 1:
+            stage += STAGE_EGRESS
+        return stage
+
+    def check_and_start(self, first: bool) -> Union[WorkerRuntime, None]:
+        """Check all worker processes and try to start failed ones.
+
+        Args:
+            first: whether the worker is tried to start at first time
+        """
+        for stage_id, worker_runtime in enumerate(self._worker_runtimes):
+            label = self._label_stage(stage_id)
+            success = worker_runtime.start(
+                first, label, self._work_path, self._shutdown, self._shutdown_notify
+            )
+            if not success:
+                return worker_runtime
+        return None
+
+
+class Controller:
+    """The manager to control Mosec process."""
+
+    def __init__(
+        self, manager: CoordinatorManager, endpoint: str, configs: Dict[str, Any]
+    ):
+        """Initialize a Mosec manager.
+
+        Args:
+            manager: manager of coordinator
+            endpoint: event of server shutdown
+            configs: config
+        """
+        self._process: Optional[subprocess.Popen] = None
+
+        self._coordinator_manager: CoordinatorManager = manager
+        self._endpoint = endpoint
+        self._server_path = Path(
+            pkg_resources.resource_filename("mosec", "bin"), "mosec"
+        )
+        self._configs: Dict[str, Any] = configs
+
+    def halt(self):
+        """Graceful shutdown."""
+        # terminate controller first and wait for a graceful period
+        if self._process is None:
+            return
+        self._process.terminate()
+        graceful_period = monotonic() + self._configs["timeout"] / 1000
+        while monotonic() < graceful_period:
+            ctr_exitcode = self._process.poll()
+            if ctr_exitcode is None:
+                sleep(0.1)
+                continue
+            # exited
+            if ctr_exitcode:  # on error
+                logger.error("mosec service halted on error [%d]", ctr_exitcode)
+            else:
+                logger.info("mosec service halted normally [%d]", ctr_exitcode)
+            break
+
+        if monotonic() > graceful_period:
+            logger.error("failed to terminate mosec service, will try to kill it")
+            self._process.kill()
+
+    def start(self):
+        """Start the Mosec process."""
+        # pylint: disable=consider-using-with
+        self._process = subprocess.Popen([self._server_path] + self._controller_args())
+        return self._process
+
+    def _controller_args(self):
+        args = []
+        self._configs.pop("dry_run")
+        self._configs["debug"] = True
```

Comment:
Debug code?

Reply:
Yes, sorry for my forgotten to remove that. Fixed.